syntax = "proto3";

// service Client_service{
//     rpc Write (WriteRequest) returns (WriteResponse) {}

//     rpc Read (ReadRequest) returns (ReadResponse) {}

//     rpc Delete (DeleteRequest) returns (DeleteResponse) {}
    
// }

service Server_service{
    rpc newReplica(newReplicaRequest) returns (newReplicaResponse) {}

    rpc WriteRequestToPrimary(WriteRequest) returns (WriteResponse) {}

    rpc DeleteRequestToPrimary(DeleteRequest) returns (DeleteResponse) {}

    rpc WriteRequestToBackup(BackupWriteRequest) returns (WriteAcknowledgement) {}

    rpc DeleteRequestToBackup(BackupDeleteRequest) returns (DeleteAcknowledgement) {}

    rpc Write (WriteRequest) returns (WriteResponse) {}

    rpc Read (ReadRequest) returns (ReadResponse) {}

    rpc Delete (DeleteRequest) returns (DeleteResponse) {}
}

message newReplicaRequest{
    Address address=1;
}

message newReplicaResponse{
    string status=1;
}


message WriteAcknowledgement{
    string status=1;
}

message DeleteAcknowledgement{
    string status=1;
}

message BackupWriteRequest{
    string name=1;
    string uuid=2;
    string content=3;
    string timestamp=4;
}

message BackupDeleteRequest{
    string uuid=1;
    string timestamp=2;
}

message WriteRequest{
    string name=1;
    string content=2;
    string uuid=3;
}

message WriteResponse{
    string status=1;
    string timestamp=2;
    string uuid=3;
}

message ReadRequest{
    string uuid=1;
}

message ReadResponse{
    string status=1;
    string name=2;
    string content=3;
    string timestamp=4;
}

message DeleteRequest{
    string uuid=1;
}

message DeleteResponse{
    string status=1;
}


message Address{
    string ip =1;
    string port =2;
}

message data_object{
    string uuid=1;
    string filename=2;
    string timestamp=3;
}

message Replica{
    string name=1;
    Address address=2;
    Address p_address=3;
    string directory=4;
    repeated data_object data_store =5;
    repeated Address backup_replicas=6;
}